### PR TITLE
[MC,COFF] Change how we handle section symbols

### PR DESCRIPTION
**Pull request by MaskRay**
13a79bbfe583e1d8cc85d241b580907260065eb8 (2017) unified `BeginSymbol` and
section symbol for ELF. This patch does the same for COFF.

* In getCOFFSection, all sections now have a `BeginSymbol` (section
  symbol). We do not need a dummy symbol name when `getBeginSymbol` is
  needed (used by AsmParser::Run and DWARF generation).
* Section symbols are in the global symbol table. `call .text` will
  reference the section symbol instead of an undefined symbol. This
  matches GNU assembler. Unlike GNU, redefining the section symbol will
  cause a "symbol 'foo0' is already defined" error (see
  `section-sym-err.s`).

### CODE REVIEW

File: llvm/lib/MC/WinCOFFObjectWriter.cpp [hunk 1]
```diff
@@ -322,27 +323,16 @@ static uint32_t getAlignment(const MCSectionCOFF &Sec) {
...
-
-  // Create a COMDAT symbol if needed.
+  // Check whether the COMDAT symbol has been reused.
   if (MCSec.getSelection() != COFF::IMAGE_COMDAT_SELECT_ASSOCIATIVE) {
     if (const MCSymbol *S = MCSec.getCOMDATSymbol()) {
-      COFFSymbol *COMDATSymbol = GetOrCreateCOFFSymbol(S);
-      if (COMDATSymbol->Section)
-        report_fatal_error("two sections have the same comdat");
-      COMDATSymbol->Section = Section;
+      if (!ComdatSymSet.insert(S->getName()).second)
```
Comment (aengelke) [inline llvm/lib/MC/WinCOFFObjectWriter.cpp:329]: getOrCreateSymbol should not create symbols with the same name. Couldn't we check this in getCOFFSection? Or store this in the symbol somehow? (I don't like the string map.)

File: llvm/lib/MC/WinCOFFObjectWriter.cpp [hunk 2]
```diff
@@ -842,12 +847,21 @@ void WinCOFFWriter::executePostLayoutBinding(MCAssembler &Asm,
...
+  auto isSectionSymbol = [](const MCSymbol &Sym) {
+    return Sym.isInSection() && Sym.getSection().getBeginSymbol() == &Sym;
+  };
+
+  for (const MCSymbol &Sym : Asm.symbols())
+    if (isSectionSymbol(Sym))
+      DefineSymbol(Sym, Asm, Layout);
+  if (Mode == DwoOnly)
+    return;
+  for (const MCSymbol &Sym : Asm.symbols())
```
Comment (aengelke) [inline llvm/lib/MC/WinCOFFObjectWriter.cpp:859]: Merge the two loops, iterating twice isn't necessary.

File: llvm/lib/MC/MCContext.cpp [hunk 3]
```diff
@@ -491,28 +491,34 @@ MCSectionMachO *MCContext::getMachOSection(StringRef Segment, StringRef Section,
...
-MCSectionELF *MCContext::createELFSectionImpl(StringRef Section, unsigned Type,
-                                              unsigned Flags,
-                                              unsigned EntrySize,
-                                              const MCSymbolELF *Group,
-                                              bool Comdat, unsigned UniqueID,
-                                              const MCSymbolELF *LinkedToSym) {
-  MCSymbolELF *R;
-  MCSymbolTableEntry &SymEntry = getSymbolTableEntry(Section);
+template <typename Symbol>
+Symbol *MCContext::getOrCreateSectionSymbol(StringRef Section) {
```
Comment (aengelke) [inline llvm/lib/MC/MCContext.cpp:495]: Move up to other getOrCreateSymbolXXX()?

**Review by aengelke — COMMENTED** [review]
Looks ok, but I'm not very familiar with COFF.

File: llvm/lib/MC/WinCOFFObjectWriter.cpp (hunk 2 above)
Comment (MaskRay) [reply llvm/lib/MC/WinCOFFObjectWriter.cpp:859]: Split DWARF places `.debug*.dwo` sections in a separate file `a.dwo`. When split DWARF is used, the first loop ensures that the  `.debug*.dwo`   symbols are present in `a.dwo`.

In addition, two loops also allow us to reduce the number of updated tests..

File: llvm/lib/MC/WinCOFFObjectWriter.cpp (hunk 1 above)
Comment (MaskRay) [reply llvm/lib/MC/WinCOFFObjectWriter.cpp:329]: I largely keep WinCOFFObjectWriter.cpp unchanged in the update.

`report_fatal_error` is kept for now, but can be moved to MCContext.cpp as a follow-up by making the symbol redefinable.

File: llvm/lib/MC/WinCOFFObjectWriter.cpp (hunk 2 above)
Comment (MaskRay) [reply llvm/lib/MC/WinCOFFObjectWriter.cpp:859]: The update reverts this change.

File: llvm/lib/MC/MCContext.cpp (hunk 3 above)
Comment (MaskRay) [reply llvm/lib/MC/MCContext.cpp]: Thx. Moved

**Review by aengelke — APPROVED** [review]
LGTM

**Pull request merged**